test/python/transpiler/test_swap_strategy.py

```diff
@@ -0,0 +1,271 @@
+        ll27_map += [[ll27[idx + 1], ll27[idx]] for idx in range(len(ll27) - 1)]
+
+        # Create a line swap strategy on this line
+        layer1 = tuple((ll27[idx], ll27[idx + 1]) for idx in range(0, len(ll27) - 1, 2))
+        layer2 = tuple((ll27[idx], ll27[idx + 1]) for idx in range(1, len(ll27), 2))
+
+        n = len(ll27)
+        for n_layers, result in [
+            (n - 4, False),
+            (n - 3, False),
+            (n - 2, True),
+            (n - 1, True),
+        ]:
+            swap_strat_ll = []
+            for idx in range(n_layers):
+                if idx % 2 == 0:
+                    swap_strat_ll.append(layer1)
+                else:
+                    swap_strat_ll.append(layer2)
+
+            strat = SwapStrategy(CouplingMap(ll27_map), tuple(swap_strat_ll))
+            self.assertEqual(strat.reaches_full_connectivity, result)
+
+    def test_possible_edges(self):
+        """Test that possible edges works as expected."""
+        coupling_map = CouplingMap(couplinglist=[(0, 1), (1, 2), (2, 3)])
+
+        strat = SwapStrategy(coupling_map, (((0, 1), (2, 3)), ((1, 2),)))
+
+        expected = set()
+        for i in range(4):
+            for j in range(4):
+                if i != j:
+                    expected.add((i, j))
+
+        self.assertSetEqual(strat.possible_edges, expected)
+
+
+class TestSwapStrategyExceptions(QiskitTestCase):
+    """A class to test the exceptions raised by swap strategies."""
+
+    def test_invalid_strategy(self):
+        """Test that a raise properly occurs."""
+
+        coupling_map = CouplingMap(couplinglist=[(0, 1), (1, 2)])
+        swap_layers = (((0, 1), (2, 3)), ((1, 2), (3, 4)))
+
+        with self.assertRaises(QiskitError):
+            SwapStrategy(coupling_map, swap_layers)
+
+    def test_invalid_line_strategy(self):
+        """Test the number of layers."""
+        message = "Negative number -1 passed for number of swap layers."
+        with self.assertRaises(ValueError, msg=message):
+            SwapStrategy.from_line([0, 1, 2], -1)
+
+
+class TestLineSwapStrategy(QiskitTestCase):
+    """A class to test the line swap strategy."""
+
+    def test_invalid_line(self):
+        """Test that lines should be longer than 1."""
+        message = "The line cannot have less than two elements, but is [1]"
+        with self.assertRaises(ValueError, msg=message):
+            SwapStrategy.from_line([1], 0)
+
+    def test_full_line(self):
+        """Test to reach full connectivity on a line."""
+
+        n_nodes = 5
+        strategy = SwapStrategy.from_line(list(range(n_nodes)))
+
+        self.assertEqual(len(strategy._swap_layers), n_nodes - 2)
+
+        # The LineSwapStrategy will apply the following permutations
+        layers = [
+            [0, 1, 2, 3, 4],  # coupling map
+            [1, 0, 3, 2, 4],  # layer 1
+            [1, 3, 0, 4, 2],  # layer 2
+            [3, 1, 4, 0, 2],  # layer 3 <-- full connectivity is reached.
+        ]
```

Comment:
Is this sufficient? Qubit 4 never reaches adjacent to the 0th node of the coupling map, so I'm not sure why it says "full connectivity is reached". 

Initially I thought a line strategy would be like figure 1 in [this paper ](https://journals.aps.org/prl/pdf/10.1103/PhysRevLett.120.110501?casa_token=-qwblnneX0QAAAAA%3A40owOyTxuxSDN-S1wmZQhHY069tDa5K3nra2nQunxxlyapGsevLgEKvfBtJUXMfR3CS-S62bquMrzQ). But maybe I don't understand what a swap strategy is in this context.

Reply:
Qubit 4 has been adjacent to every other qubit when layer 3 is reached. See here for more details: https://arxiv.org/abs/2202.03459

Reply:
Ok yes it's the same odd-even steps as that paper I linked, but without the last step which I see is not necessary.